src/test/java/io/vertx/core/ContextTest.java

```diff
@@ -514,29 +516,38 @@ private void checkDuplicate(ContextInternal ctx, ContextInternal duplicated) thr
 
   @Test
   public void testDuplicateWorkerConcurrency() throws Exception {
-    ContextInternal ctx = createWorkerContext();
-    ContextInternal dup1 = ctx.duplicate();
-    ContextInternal dup2 = ctx.duplicate();
-    CyclicBarrier barrier = new CyclicBarrier(3);
-    dup1.runOnContext(v -> {
-      assertTrue(Context.isOnWorkerThread());
-      try {
-        barrier.await(10, TimeUnit.SECONDS);
-      } catch (Exception e) {
-        fail(e);
-      }
-    });
-    dup2.runOnContext(v -> {
-      assertTrue(Context.isOnWorkerThread());
-      try {
-        barrier.await(10, TimeUnit.SECONDS);
-      } catch (Exception e) {
-        fail(e);
-      }
-    });
-    barrier.await(10, TimeUnit.SECONDS);
+    testDuplicateWorkerConcurrency((ctx, task) -> ctx.runOnContext(v -> task.run()));
+    testDuplicateWorkerConcurrency((ctx, task) -> ctx.execute(v -> task.run()));
+    testDuplicateWorkerConcurrency((ctx, task) -> ctx.execute(null, v -> task.run()));
+    testDuplicateWorkerConcurrency(ContextInternal::execute);
+    testDuplicateWorkerConcurrency((ctx, task) -> ctx.emit(v -> task.run()));
+    testDuplicateWorkerConcurrency((ctx, task) -> ctx.emit(null, v -> task.run()));
+  }
+
+  private void testDuplicateWorkerConcurrency(BiConsumer<ContextInternal, Runnable> task) throws Exception {
+    ContextInternal worker = ((VertxInternal)vertx).createWorkerContext();
+    ContextInternal[] contexts = new ContextInternal[] { worker.duplicate(), worker.duplicate()};
+    waitFor(contexts.length);
+    AtomicBoolean owner = new AtomicBoolean();
+    CountDownLatch latch = new CountDownLatch(contexts.length);
+    for (ContextInternal context : contexts) {
+      task.accept(context, () -> {
+        assertFalse(owner.get());
```

Comment:
There is a chance 2 threads executing this and none will be the owner, you need to make an atomic comparison:
```
assertTrue(owner.compareAndSet(false, true));
```
then you don't need to set owner to true on after the `try`

Reply:
good catch, if there is no bug then it is ok, but if that's buggy (which is what we are testing here) indeed it is racy